keps/sig-api-machinery/customresources-pruning-and-defaulting.md

```diff
@@ -0,0 +1,486 @@
+
+Clearly, we can make the algorithm deterministic by choosing one, e.g. the first matching branch. This choice is arbitrary, hard to understand, and gets even more confusing with even more nested `allOf`, `anyOf` or even `not`.
+
+We can later become less restrictive about the allowed positions of default values in the scheme. I.e. the restriction of only allowing default outside of `allOf`, `anyOf`, `oneOf` and `not` is a starting point with a way forward if necessary.
+
+### Mixing of Schema and Value Validation
+
+The OpenAPI validation schema mixes the actual structural schema validation (which is usually done by the Golang struct JSON decoding for native types) 
+and the value validation (usually done in the validation step of the API server handler pipeline for native types). 
+
+For CRDs we cannot distinguish both. This was first noticed by @lavalamp in https://github.com/kubernetes/kubernetes/pull/64907#issuecomment-397015030.
+
+#### Example 6
+
+> Assume 
+>
+>```json
+>{
+>   "properties": {
+>       "a": {"type": "string", "pattern": "<some-long-regex-for-ips>", "format": "ip"},
+>       "b": {"properties": {"x": {}, "y": {}}}
+>   }
+>}
+>```
+>
+> The type and properties are usually called schema validation, while regex patterns and formats are about values. The latter would be tested in the validation phase, the former would be checked during decoding for native types in the JSON decoder.
+
+**Remark:** the line between type and format is blurry. Format is optional to be processed by tools, and the value is open ended. In Kube we have some formats (like `date`) which correspond to custom JSON unmarshallers in the native types. That format would be verified during decoding, although technically a date is just a string. So we might want to replicate that behaviour at some point.
+
+For pruning and defaulting we have to apply the OpenAPI validation schema inside of the decoder step (the left box inside apiextensions-apiserver in the figure above). This is considerably earlier than for native types. This has a number of implications:
+* We have to apply full OpenAPI value validation (e.g. regular expressions, propositional evaluation) during decoding.
+* Our generic registry applies validation after defaulting. We need the other way around for CRDs.
+* Our generic registry expects defaulting to always succeed (no error result type). CRD validation needed by defaulting can fail.
+
+To avoid these and to get sane semantics, **we propose to split the CRD validation into two steps:**
+1. During decoding we validate using a skeleton schema which lacks value validations and propositional logic (`anyOf`, `allOf`, `oneOf`, `not`).
+2. During defaulting we validate using the skeleton schema and apply the resulting defaulter.
+During standard generic registry validation phase we validate using the full validation schema.
```

Comment:
Yes, we will. We have this hardcoded in the mutating admission webhook code (https://github.com/kubernetes/kubernetes/blob/e4c219df42c77ecb8f0588197072bef81bca7429/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go#L152) and this will be called the same way for CRDs.

Reply:
added a note